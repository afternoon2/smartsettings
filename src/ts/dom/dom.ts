import { html } from 'lighterhtml';
import { PanelOptions } from '../panel/panel.types';

export function isValidElement(parent: HTMLElement | string): boolean {
  if (typeof parent === 'string') {
    return Boolean(document.querySelector(parent));
  }
  if (parent instanceof HTMLElement && document.body.contains(parent)) {
    return true;
  }
  return false;
}

export function getParentElement(parent: HTMLElement | string): HTMLElement {
  if (parent instanceof HTMLElement) {
    return parent;
  }
  return <HTMLElement>document.querySelector(parent);
}

export function createPanelNode(
  title: string,
  id: string,
  options: PanelOptions = {},
): HTMLDivElement {
  const rootNode: HTMLDivElement = document.createElement('div');
  rootNode.classList.add('panel');
  if (options.locked) {
    rootNode.classList.add('locked');
  }
  if (options.invisible) {
    rootNode.classList.add('hidden');
  }
  rootNode.setAttribute('id', id);
  const template: HTMLElement = html`
    <header class="panel__header">
      <a class="panel__dragLink ${options.draggable ? '' : 'hidden'}">
        ✣
      </a>
      <p class="panel__title" title="${title}">
        ${title}
      </p>
      <a class="panel__collapseLink">
        ${options.collapsed ? '▷' : '▽'}
      </a>
    </header>
    <main class="panel__main${options.collapsed ? ' hidden' : ''}">
    </main>
  `;
  rootNode.appendChild(template);
  return rootNode;
}

