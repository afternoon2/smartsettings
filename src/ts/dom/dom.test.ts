import { isValidElement, createPanelNode, getParentElement } from './dom';

describe('panel.dom', () => {
  describe('checkParentElement', () => {
    test('Checking string parent element', () => {
      const validSelector = '.parent';
      const invalidSelector = 'parent';
      const parent = document.createElement('div');
      parent.classList.add('parent');
      document.body.appendChild(parent);
      expect(isValidElement(validSelector)).toBe(true);
      expect(isValidElement(invalidSelector)).toBe(false);
    });

    test('Checking html parent element', () => {
      const parent = document.createElement('div');
      parent.classList.add('parent');
      expect(isValidElement(parent)).toBe(false);
      document.body.appendChild(parent);
      expect(isValidElement(parent)).toBe(true);
    });
  });

  describe('getParentElement', () => {
    test('Get parent element from string', () => {
      const parent = document.createElement('div');
      parent.classList.add('parent');
      document.body.appendChild(parent);
      const gotParent = getParentElement('.parent');
      expect(gotParent).toMatchObject(parent);
    });

    test('Get parent element from HTMLElement', () => {
      const parent = document.createElement('div');
      parent.classList.add('parent');
      document.body.appendChild(parent);
      const gotParent = getParentElement(parent);
      expect(gotParent).toMatchObject(parent);
    });
  });

  describe('createPanelNode', () => {
    test('Create panel node with required parameters', () => {
      const id = 'test_id';
      const titleStr = 'Test title';
      const node = createPanelNode(titleStr, id);
      const dragLink = <HTMLLinkElement>node.querySelector('.panel__dragLink');
      const title = <HTMLParagraphElement>node.querySelector('.panel__title');
      const collapseLink = <HTMLLinkElement>node.querySelector('.panel__collapseLink');
      const main = <HTMLElement>node.querySelector('.panel__main');
      expect(node.classList.contains('locked')).toBe(false);
      expect(node).toBeInstanceOf(HTMLDivElement);
      expect(dragLink.classList.contains('hidden')).toBe(true);
      expect((<string>title.textContent).search('Test title')).toBeGreaterThan(-1);
      expect((<string>collapseLink.textContent).search('▽')).toBeGreaterThan(-1);
      expect(main.classList[1]).toBe(undefined);
      expect(node).toMatchSnapshot();
    });

    test('Create panel node with not required options', () => {
      const options = {
        collapsed: true,
        draggable: true,
        locked: true,
        invisible: true,
      };
      const node = createPanelNode('Test title', 'test_id', options);
      const dragLink = <HTMLLinkElement>node.querySelector('.panel__dragLink');
      const title = <HTMLParagraphElement>node.querySelector('.panel__title');
      const collapseLink = <HTMLLinkElement>node.querySelector('.panel__collapseLink');
      const main = <HTMLElement>node.querySelector('.panel__main');
      expect(node.classList.contains('locked')).toBe(true);
      expect(node.classList.contains('hidden')).toBe(true);
      expect(node).toBeInstanceOf(HTMLDivElement);
      expect(dragLink.classList.contains('hidden')).toBe(false);
      expect((<string>title.textContent).search('Test title')).toBeGreaterThan(-1);
      expect((<string>collapseLink.textContent).search('▷')).toBeGreaterThan(-1);
      expect(main.classList.contains('hidden')).toBe(true);
      expect(node).toMatchSnapshot();
    });
  });
});
