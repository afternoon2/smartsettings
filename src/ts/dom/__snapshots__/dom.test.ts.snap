// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`panel.dom createPanelNode Create panel node with not required options 1`] = `
<div
  class="panel locked hidden"
  id="test_id"
>
  <header
    class="panel__header"
  >
    
      
    <a
      class="panel__dragLink "
    >
      
        ✣
      
    </a>
    
      
    <p
      class="panel__title"
      title="Test title"
    >
      
        
      Test title
      <!---0.411551%-->
      
      
    </p>
    
      
    <a
      class="panel__collapseLink"
    >
      
        
      ▷
      <!---0.411551%-->
      
      
    </a>
    
    
  </header>
  <main
    class="panel__main hidden"
  >
    
    
  </main>
</div>
`;

exports[`panel.dom createPanelNode Create panel node with required parameters 1`] = `
<div
  class="panel"
  id="test_id"
>
  <header
    class="panel__header"
  >
    
      
    <a
      class="panel__dragLink hidden"
    >
      
        ✣
      
    </a>
    
      
    <p
      class="panel__title"
      title="Test title"
    >
      
        
      Test title
      <!---0.411551%-->
      
      
    </p>
    
      
    <a
      class="panel__collapseLink"
    >
      
        
      ▽
      <!---0.411551%-->
      
      
    </a>
    
    
  </header>
  <main
    class="panel__main"
  >
    
    
  </main>
</div>
`;
